File: exercise_01/mutex_bridge/mutex_bridge.cpp
#include <iostream>
#include <string>
#include <thread>
#include <chrono>
#include <mutex>

using namespace std;

mutex bridgeMutex;

bool isGoingLeft, isGoingRight;
string bridge;

string getDirection()
{
  return isGoingLeft ? "left" : "right";
}

void cross()
{
  string direction = getDirection();

  cout << "=================================" << endl;
  cout << "🚗 Car on the bridge is going " << direction << endl;

  this_thread::sleep_for(chrono::seconds(5));

  cout << "✅ Car finished going to the " << direction << endl;
  cout << "---------------------------------" << endl;
}

void tryCrossingBridge(string s)
{
  if (s == "left")
  {
    isGoingLeft = true;
    cross();
    isGoingLeft = false;
  }
  if (s == "right")
  {
    isGoingRight = true;
    cross();
    isGoingRight = false;
  }
}

void keepTryingCrossingBrigeToThe(string direction, int numberSecondsSleeping) {
  while(true) {
    cout << "🕑 Car is waiting to go to the " << direction << endl;
    this_thread::sleep_for(chrono::seconds(numberSecondsSleeping));

    bridgeMutex.lock();

    tryCrossingBridge(direction);

    bridgeMutex.unlock();
  }
}

void bridgeWatcher()
{
  while (true)
  {
    if (isGoingLeft && isGoingRight)
    {
      cout << "💥 CRASH HAPPENED!" << endl;
      throw 255;
    }
  }
}

int main()
{
  thread first(keepTryingCrossingBrigeToThe, "left", 2);
  thread second(keepTryingCrossingBrigeToThe, "right", 3);
  thread third(bridgeWatcher);

  while (true)
    ;

  return 0;
}

File: exercise_01/mutex_bridge/mutex_bridge.go
package main

import (
	"fmt"
	"os"
	"sync"
	"time"
)

var mutex = &sync.Mutex{} // mutual-exclusion lock
var bridge string

var isGoingLeft bool
var isGoingRight bool

func keepTryingCrossingBridgeToThe(direction string, n time.Duration) {
	for {
		fmt.Println("🕑", "Car is waiting to go to the", direction)
		time.Sleep(n * time.Second)

		mutex.Lock()

		tryCrossingBridge(direction)

		mutex.Unlock()
	}
}

func tryCrossingBridge(s string) {
	switch s {
	case "left":
		isGoingLeft = true
		cross()
		isGoingLeft = false
		break
	case "right":
		isGoingRight = true
		cross()
		isGoingRight = false
		break
	}
}

func cross() {
	var direction string = getDirection()

	fmt.Println("=================================")
	fmt.Println("🚗", "Car on the bridge is going", direction)

	time.Sleep(5 * time.Second)

	fmt.Println("✅", "Car finished going to the", direction)
	fmt.Println("---------------------------------")
}

func getDirection() string {
	if isGoingLeft {
		return "left"
	} else {
		return "right"
	}
}

func bridgeWatcher() {
	for {
		if isGoingLeft && isGoingRight {
			fmt.Println("💥", "CRASH HAPPENED!")
			os.Exit(-1) // all 1s in binary
		}
	}
}

func main() {
	go keepTryingCrossingBridgeToThe("left", 2)
	go keepTryingCrossingBridgeToThe("right", 3)

	go bridgeWatcher()

	fmt.Scanln()
}
